src/facade/dragonfly_connection.cc

```diff
@@ -190,35 +191,32 @@ size_t Connection::PipelineMessage::StorageCapacity() const {
   return storage.capacity() + args.capacity();
 }
 
-template <class... Ts> struct Overloaded : Ts... {
-  using Ts::operator()...;
-
-  template <typename T, typename D> size_t operator()(const unique_ptr<T, D>& ptr) {
-    return operator()(*ptr.get());
-  }
-};
-
-template <class... Ts> Overloaded(Ts...) -> Overloaded<Ts...>;
-
 size_t Connection::MessageHandle::UsedMemory() const {
-  // TODO: don't count inline size
-  auto pub_size = [](const PubMessage& msg) -> size_t {
-    return sizeof(PubMessage) + (msg.channel_len + msg.message_len);
-  };
-  auto msg_size = [](const PipelineMessage& arg) -> size_t {
-    return sizeof(PipelineMessage) + arg.args.capacity() * sizeof(MutableSlice) +
-           arg.storage.capacity();
-  };
-  auto monitor_size = [](const MonitorMessage& arg) -> size_t { return arg.capacity(); };
-  auto acl_update_size = [](const AclUpdateMessage& msg) -> size_t {
-    return sizeof(AclUpdateMessage);
-  };
-  auto migration_request_size = [](const MigrationRequestMessage& msg) -> size_t {
-    return sizeof(MigrationRequestMessage);
+  struct MessageSize {
+    size_t operator()(const PubMessagePtr& msg) {
+      return sizeof(PubMessage) + (msg->channel_len + msg->message_len);
+    }
+    size_t operator()(const PipelineMessagePtr& msg) {
+      return sizeof(PipelineMessage) + msg->args.capacity() * sizeof(MutableSlice) +
+             msg->storage.capacity();
+    }
+    size_t operator()(const MonitorMessage& msg) {
+      return msg.capacity();
+    }
+    size_t operator()(const AclUpdateMessage& msg) {
+      return 0;
+    }
+    size_t operator()(const MigrationRequestMessage& msg) {
+      return 0;
+    }
```

Comment:
Intentional: this is the heap size _only_ (should add a comment), because the size of the variant itself is included below